src/tss2-tcti/tcti-device.c

```diff
@@ -110,7 +110,25 @@ tcti_device_transmit (
     tcti_common->state = TCTI_STATE_RECEIVE;
     return TSS2_RC_SUCCESS;
 }
-
+/*
+ * This receive function deviates from the spec a bit. Calling this function
+ * with a NULL 'tctiContext' parameter *should* result in the required size for
+ * the response buffer being returned to the caller. We would typically do this
+ * by reading the response's header and then returning the size to the caller.
+ * We can't do that on account of the TPM2 kernel driver closing any connection
+ * that doesn't read the whole response buffer in one 'read' call.
+ *
+ * Instead, if the caller queries the size, we return 4k just to be on the
+ * safe side. We do *not* however verify that the provided buffer is large
+ * enough to hold the full response (we can't). If the caller provides us with
+ * a buffer less than 4k we'll read as much of the response as we can given
+ * the size of the buffer. If we get enough of the response to read the size
+ * field, we check to see if the buffer was large enough to get the full
+ * response. If the response header claims it's larger than the provided
+ * buffer we print a warning. This allows "expert applications" to
+ * precalculate the required response buffer size for whatever commands they
+ * may send.
+ */
```

Comment:
As mentioned in the issue tracking this, should we fix the driver instead? Is this just an implementation detail of the tpm driver or a constraint of the TPM? If the former, I really think we should change the driver instead of having workarounds in user-space.

Reply:
@flihp @tstruk @AndreasFuchsSIT ^

Reply:
It would be great if interaction with the kernel driver could be made a bit more friendly to user space. Having the time to hack up a patch for the kernel is another issue. For now, userspace needs this workaround.

Reply:
@flihp fair enough.